### PR TITLE
fix(lsp): update document dependencies on configuration change

### LINKED ISSUE #17555: lsp - Updating import map requires lsp restart or code change
It seems that after updating an import map, you need to restart the language server to get the changes reflected or cause a code change to the file whose diagnostics you want updated.

### PR DESCRIPTION
**Pull request by dsherret**
Closes #17555

Video shows current behaviour on main then behaviour on this branch:

https://user-images.githubusercontent.com/1609021/215010703-fbb1d8a1-d93d-49d1-9e6f-a3490e3361db.mp4

### CODE REVIEW

File: cli/lsp/documents.rs [hunk 1]
```diff
@@ -290,28 +291,27 @@ impl Document {
...
       &specifier,
-      content.clone(),
-      maybe_headers,
+      text_info.text(),
+      maybe_headers.as_ref(),
       maybe_resolver,
     );
     let dependencies =
       Arc::new(DocumentDependencies::from_maybe_module(&maybe_module));
-    // todo(dsherret): retrieve this from the parsed source if it exists
```
Comment (dsherret) [inline cli/lsp/documents.rs:308]: Not relevant anymore. We don't keep the parsed sources in memory.

File: cli/lsp/documents.rs [hunk 2]
```diff
@@ -1112,7 +1137,6 @@ impl Documents {
     maybe_import_map: Option<Arc<import_map::ImportMap>>,
     maybe_config_file: Option<&ConfigFile>,
   ) {
-    // TODO(@kitsonk) update resolved dependencies?
```
Comment (dsherret) [inline cli/lsp/documents.rs:1115]: This PR resolves this todo.

File: cli/tests/integration/lsp_tests.rs [hunk 3]
```diff
@@ -508,15 +508,11 @@ fn lsp_import_map_config_file() {
...
-  fs::write(
-    temp_dir.path().join("lib").join("b.ts"),
-    r#"export const b = "b";"#,
-  )
-  .unwrap();
+  let import_map_text =
+    serde_json::to_string_pretty(&load_fixture("import-map.json")).unwrap();
+  temp_dir.write("import-map.json", import_map_text);
+  temp_dir.create_dir_all("lib");
+  temp_dir.write("lib/b.ts", r#"export const b = "b";"#);
```
Comment (dsherret) [inline cli/tests/integration/lsp_tests.rs:515]: I tried to add an integration test and it was too difficult, so I just added one in Documents.

File: cli/lsp/documents.rs (hunk 1 above)
Comment (dsherret) [reply cli/lsp/documents.rs:308]: Wait, I'm wrong here. We do in maybe_parsed_source. Will look into this

File: cli/lsp/documents.rs [hunk 4]
```diff
@@ -322,6 +324,19 @@ impl Document {
     }))
   }
 
+  fn with_new_resolver(
+    &self,
+    maybe_resolver: Option<&dyn deno_graph::source::Resolver>,
+  ) -> Self {
+    Self::new(
```
Comment (dsherret) [inline cli/lsp/documents.rs:331]: This is wrong. I will open up a fix in a bit.

File: cli/lsp/documents.rs (hunk 1 above)
Comment (dsherret) [reply cli/lsp/documents.rs:308]: It should use less memory now because the parsed source and document will share the same SourceTextInfo.

Comment (bartlomieju) [reply cli/lsp/documents.rs:308]: Very nice 🎉

**Review by bartlomieju — APPROVED** [review]
LGTM, great fix

**Pull request merged**